Add draft github pull-request template

leifdenby opened the pull request:
The aim of this PR is to add a pull-request template for use on github when add pull-requests to the common repo. The aim of my initial commit is to start the discussion on how we would like pull-requests to progress when collaborating on neural-lam.

As you will see from the template I have created so far the aim is to:

- create placeholders for the PR creator to add context so we know what the PR is for
- detail the steps of getting the PR reviewed and merged into the repo

Looking forward to hearing your thoughts all :)

Here's a direct link so you can see it rendered: https://github.com/leifdenby/neural-lam/blob/maint/add-github-pr-template/.github/pull_request_template.md

sadamov commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,41 @@
+
+## Type of change
+
+- [ ] Bug fix (non-breaking change which fixes an issue)
+- [ ] New feature (non-breaking change which adds functionality)
+- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
```
```suggestion
- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [ ] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
```

leifdenby replied:
I love both additions, thank you!

joeloskarsson replied:
The documentation entry was not added correctly now.

sadamov reviewed (commented):
Thanks, this will prove very useful! I added some smaller discussion points.

SimonKamuk commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,41 @@
+
+- [ ] My branch is up-to-date with the target branch - if not update your fork with the changes from the target branch (use `pull` with `--rebase` option if possible).
+- [ ] I have performed a self-review of my code
+- [ ] I have commented my code, particularly in hard-to-understand areas
```
I completely agree with the added README suggestion, but I am not entirely convinced we should discourage the use of inline comments. Maybe the text could be changed to something like 
"I have commented my code where necessary, for example in hard-to-understand areas"

leifdenby replied:
How about we write:

```markdown
- [ ] I have added docstrings to any new/modified functions
- [ ] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [ ] I have updated the [README](README.MD) to cover introduced code changes
```

joeloskarsson replied:
I think that is good. Test (linting) should catch missing docstrings right? So that is maybe superfluous if test pass?

leifdenby replied:
> Test (linting) should catch missing docstrings right? So that is maybe superfluous if tests pass?

True, the linting checks that docstrings have been written, but not that they are useful. Maybe we should leave this in to just remind people to think through the docstrings?

joeloskarsson replied:
Yes, I think that's good. Proposal:
```suggestion
- [ ] For any new/modified functions/classes I have added docstrings that clearly describe its purpose, expected inputs and returned values
- [ ] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [ ] I have updated the [README](README.MD) to cover introduced code changes
```

joeloskarsson replied:
Update the text about doc-strings

SimonKamuk commented:
Great idea with the template, and good suggestions by @sadamov. I have just added a small comment above, and have no further changes - great work!

sadamov commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,41 @@
+< Please also include relevant motivation and context. >
+
+< List any dependencies that are required for this change. >
+
```
This is actually not required, as it can be specified directly in the side-menu

leifdenby replied:
Ok, we can put this in. We can always revise this template later if we don't use parts of it

joeloskarsson commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,41 @@
+- [ ] I have commented my code, particularly in hard-to-understand areas
+- [ ] I have added tests that prove my fix is effective or that my feature works
+- [ ] I have requested a reviewer and an assignee (assignee is responsible for merging)
+
```
```suggestion
- [ ] I have given the PR a name that clearly describes the changes, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form).
```

My idea here is that we should be able to use the PR title directly as the commit message when merging (so the assignee does not have to spend extra time formulating a descriptive commit message when merging). I think it's nice to follow the convention of imperative mood for commit messages.

joeloskarsson commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,41 @@
+  reflecting type of change, for example "bug fixes", add section where
+  missing)
+
+## Checklist for assignee
+
+- [ ] PR is up to date with the base branch
+- [ ] the tests passing
+- [ ] author has added an entry to the changelog (and designated the change as *added*, *changed* or *fixed*)
```
I view this similarly as the checklist for reviewers as not really needed in this template.

leifdenby commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,41 @@
+- [ ] Bug fix (non-breaking change which fixes an issue)
+- [ ] New feature (non-breaking change which adds functionality)
+- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
+
```
Would actually err on the side not needing this. If a file-by-file description is needed then I would suggest the PR is trying to do too much

joeloskarsson replied:
I'd be fine with leaving it out for that reason

leifdenby commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,41 @@
+## Checklist for reviewers
+
+Each PR comes with its own improvements and flaws. The reviewer should check the following:
+- [ ] the code readable
+- [ ] the code well tested
+- [ ] the code documented
+- [ ] the code easy to maintain
```
we can also leave "easy to maintain" out? The other three cover this some degree and this isn't explicit enough about how we measure this that the comment is probably superfluous

leifdenby replied:
I will remove "easy to maintain" since that is hard to measure objectively, we can always adapt this template later if we find it is missing something or doesn't quite work :)

leifdenby commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,41 @@
+- [ ] I have added tests that prove my fix is effective or that my feature works
+- [ ] I have requested a reviewer and an assignee (assignee is responsible for merging)
+
+## Checklist for reviewers
```
Yes, these steps spell out the review process, but I do think it is useful to see how far along the review process a given PR is. And also, as you say, reviewing guidelines should go somewhere. Putting the in the PR ensures that people have them when they need them too, and are more likely to use them too. But maybe this points to the need to have a brief discussion about the review process and then we can decide where to put it? Having them in the PR template has worked well for me before, but I am also happy to discussing doing this differently :)

leifdenby commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,41 @@
+- [ ] the code documented
+- [ ] the code easy to maintain
+
+## Author checklist after completed review
+
+- [ ] I have added a line to the CHANGELOG describing this change (in section
+  reflecting type of change, for example "bug fixes", add section where
+  missing)
```
> I guess a drawback with just editing the changelog immediately is that what should be written there could change after discussions in the PR, but I think that is quite rare, and then you just have to remember to update the changelog edit as well.

Yes, to me this is the key drawback. Sometimes people will want to fix/change/add something and they way they have achieved that might change as the review process progresses. In my experience this is often the case. Also, people often forget to update the changelog so this is a a little last-minute reminder to do that before merging. And finally, a given PR doesn't have a URL until it has been created, and so updating the changelog with a separate commit after the PR as been created is typically necessary.

leifdenby replied:
> A question about the changelog sections: This here seems to apply that one should add arbitrary new section if there is not a new one? But below the changelog sections are described as fixed (_added_, _changed_, _fixed_). What is the general idea for this? If we are sticking to explicitly the sections listed at https://keepachangelog.com/en/1.1.0/, then I think we should make that clear also here. One way to do that is to always have those sections under _unreleased_, and then there is never a need to add sections with a PR. Empty sections at the time of a release can just be removed.

Yes, I think having fixed sections sounds like a good idea. I was going to suggest we discuss what these should be so that we are on the same page about it. One section that might be good to include (in addition to *added, changed, fixed*) is *maintenance* (these are things that don't change fhe functionality of the code at all, but are say necessary package updates because something breaks, fixing the ci/cd linting, testing etc update). We can also put those in the other sections since they will often *fix* something, but sometimes this info isn't relevant to you if you just want to decide whether to update or not.

joeloskarsson replied:
I'm ok with keeping it like this then. Just make the changelog categories explicit and consistent in the template, _added, changed, fixed, maintenance_ sound good to me.

leifdenby commented:
I am going to promote this to an actual PR from a draft PR now the discussion is well underway :smile:

joeloskarsson commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,42 @@
+- [ ] I have commented my code, particularly in hard-to-understand areas
+- [ ] I have added tests that prove my fix is effective or that my feature works
+- [ ] I have requested a reviewer and an assignee (assignee is responsible for merging)
+- [ ] I have given the PR a name that clearly describes the change, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form).
```
There's a missing )
```suggestion
- [ ] I have given the PR a name that clearly describes the change, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form)).
```

joeloskarsson commented:
Any updates on this @leifdenby? Are there specific points you want more input on? Out of my points it is mainly the checkbox for adding tests that requires some further discussion. It would be good to try to merge this soon.

leifdenby commented:
TODO

- [ ] remove line about having added tests
- [ ] add option for documentation under the type of change
- [ ] update changelog sections to choose from to be explicitly listed as *added, changed, fixed, maintenance*

leifdenby commented:
Ok, I think this is ready to go now @joeloskarsson. Do you mind giving it a last read over?

joeloskarsson commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,50 @@
+## Checklist for reviewers
+
+Each PR comes with its own improvements and flaws. The reviewer should check the following:
+- [ ] the code readable
+- [ ] the code well tested
+- [ ] the code documented (including return types and parameters)
+- [ ] the code easy to maintain
```
```suggestion
- [ ] the code is readable
- [ ] the code is well tested
- [ ] the code is documented (including return types and parameters)
- [ ] the code is easy to maintain
```

joeloskarsson commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,50 @@
+## Checklist for assignee
+
+- [ ] PR is up to date with the base branch
+- [ ] the tests passing
```
```suggestion
- [ ] the tests pass
```

joeloskarsson reviewed (changes requested):
Gave it a proofread and fixed some grammar. After these fixes + updating the docstring text this should be good to merge!

leifdenby commented:
Thanks for the thorough review @joeloskarsson! I will merge this in

joeloskarsson reviewed (approved):
Great, all looking good now!
